==========> Minecraft.java
--- Minecraft.java
+++ Minecraft.java
@@ -121,0 +121,2 @@
+import com.minecrafttas.tasmodog.container.Playback;
+
@@ -129,1 +131,1 @@
-    private static Minecraft theMinecraft;
+    public static Minecraft theMinecraft;
@@ -136,1 +138,1 @@
-    private Timer timer;
+    public Timer timer;
@@ -164,1 +166,1 @@
-    protected MinecraftApplet mcApplet;
+    public MinecraftApplet mcApplet;
@@ -180,1 +182,1 @@
-    private static File minecraftDir = null;
+    public static File minecraftDir = null;
@@ -602,1 +604,1 @@
-                System.exit(0);
+//                System.exit(0);
@@ -621,0 +623,1 @@
+        com.minecrafttas.tasmodog.TASmod.instance.init(this); // initialize tasmod
@@ -681,0 +684,1 @@
+        com.minecrafttas.tasmodog.TASmod.instance.render(); // Call TASmod Render/Gameloop Tick before runTick() to be able to hijack that
@@ -712,1 +716,1 @@
-        if(!Keyboard.isKeyDown(65))
+        if(!com.minecrafttas.tasmodog.virtual.VirtualKeyboard.isKeyDown(65))
@@ -734,4 +738,4 @@
-        if(!Display.isActive() && fullscreen)
-        {
-            toggleFullscreen();
-        }
+//        if(!Display.isActive() && fullscreen)
+//        {
+//            toggleFullscreen();
+//        }
@@ -755,1 +759,1 @@
-        if(Keyboard.isKeyDown(65))
+        if(com.minecrafttas.tasmodog.virtual.VirtualKeyboard.isKeyDown(65))
@@ -813,1 +817,1 @@
-        if(Keyboard.isKeyDown(60))
+        if(com.minecrafttas.tasmodog.virtual.VirtualKeyboard.isKey60Down)
@@ -1023,4 +1027,4 @@
-        if(!Display.isActive())
-        {
-            return;
-        }
+//        if(!Display.isActive())
+//        {
+//            return;
+//        }
@@ -1316,0 +1320,24 @@
+        // Moved from EntityRenderer.java / basically just hooked up.
+        if(inGameHasFocus) {
+            
+            float f1 = gameSettings.mouseSensitivity * 0.6F + 0.2F;
+            float f2 = f1 * f1 * f1 * 8F;
+            
+            mouseHelper.mouseXYChange(); // Call the real one, but actually call the virtual one
+            
+            thePlayer.setAngles((float) mouseHelper.deltaX * f2, (float) mouseHelper.deltaY * f2);
+            
+            // interpolation part 1
+        	if (com.minecrafttas.tasmodog.TASmod.instance.getInputContainer() instanceof Playback && !((Playback) com.minecrafttas.tasmodog.TASmod.instance.getInputContainer()).isVisible()) {
+        		com.minecrafttas.tasmodog.virtual.VirtualCamera.prevRotationPitch = com.minecrafttas.tasmodog.virtual.VirtualCamera.rotationPitch;
+        		com.minecrafttas.tasmodog.virtual.VirtualCamera.prevRotationYaw = com.minecrafttas.tasmodog.virtual.VirtualCamera.rotationYaw;
+        	} else {        		
+        		com.minecrafttas.tasmodog.virtual.VirtualCamera.prevRotationPitch = thePlayer.prevRotationPitch;
+        		com.minecrafttas.tasmodog.virtual.VirtualCamera.prevRotationYaw = thePlayer.prevRotationYaw;
+        	}
+        	com.minecrafttas.tasmodog.virtual.VirtualCamera.rotationPitch = thePlayer.rotationPitch;
+        	com.minecrafttas.tasmodog.virtual.VirtualCamera.rotationYaw = thePlayer.rotationYaw;
+        }
+        com.minecrafttas.tasmodog.virtual.VirtualCamera.lastX = com.minecrafttas.tasmodog.virtual.VirtualMouse.getEventX();
+        com.minecrafttas.tasmodog.virtual.VirtualCamera.lastY = com.minecrafttas.tasmodog.virtual.VirtualMouse.getEventY();
+        // End-Edit
@@ -1321,1 +1349,1 @@
-                if(!Mouse.next())
+                if(!com.minecrafttas.tasmodog.virtual.VirtualMouse.next())
@@ -1325,2 +1353,2 @@
-                KeyBinding.setKeyBindState(Mouse.getEventButton() - 100, Mouse.getEventButtonState());
-                if(Mouse.getEventButtonState())
+                KeyBinding.setKeyBindState(com.minecrafttas.tasmodog.virtual.VirtualMouse.getEventButton() - 100, com.minecrafttas.tasmodog.virtual.VirtualMouse.getEventButtonState());
+                if(com.minecrafttas.tasmodog.virtual.VirtualMouse.getEventButtonState())
@@ -1328,1 +1356,1 @@
-                    KeyBinding.onTick(Mouse.getEventButton() - 100);
+                    KeyBinding.onTick(com.minecrafttas.tasmodog.virtual.VirtualMouse.getEventButton() - 100);
@@ -1333,1 +1361,1 @@
-                    int i1 = Mouse.getEventDWheel();
+                    int i1 = com.minecrafttas.tasmodog.virtual.VirtualMouse.getEventDWheel();
@@ -1352,1 +1380,1 @@
-                        if(!inGameHasFocus && Mouse.getEventButtonState())
+                        if(!inGameHasFocus && com.minecrafttas.tasmodog.virtual.VirtualMouse.getEventButtonState())
@@ -1370,1 +1398,1 @@
-                if(!Keyboard.next())
+                if(!com.minecrafttas.tasmodog.virtual.VirtualKeyboard.next())
@@ -1374,2 +1402,2 @@
-                KeyBinding.setKeyBindState(Keyboard.getEventKey(), Keyboard.getEventKeyState());
-                if(Keyboard.getEventKeyState())
+                KeyBinding.setKeyBindState(com.minecrafttas.tasmodog.virtual.VirtualKeyboard.getEventKey(), com.minecrafttas.tasmodog.virtual.VirtualKeyboard.getEventKeyState());
+                if(com.minecrafttas.tasmodog.virtual.VirtualKeyboard.getEventKeyState())
@@ -1377,1 +1405,1 @@
-                    KeyBinding.onTick(Keyboard.getEventKey());
+                    KeyBinding.onTick(com.minecrafttas.tasmodog.virtual.VirtualKeyboard.getEventKey());
@@ -1379,1 +1407,1 @@
-                if(Keyboard.getEventKeyState())
+                if(com.minecrafttas.tasmodog.virtual.VirtualKeyboard.getEventKeyState())
@@ -1381,1 +1409,1 @@
-                    if(Keyboard.getEventKey() == 87)
+                    if(com.minecrafttas.tasmodog.virtual.VirtualKeyboard.getEventKey() == 87)
@@ -1391,1 +1419,1 @@
-                            if(Keyboard.getEventKey() == 1)
+                            if(com.minecrafttas.tasmodog.virtual.VirtualKeyboard.getEventKey() == 1)
@@ -1395,1 +1423,1 @@
-                            if(Keyboard.getEventKey() == 31 && Keyboard.isKeyDown(61))
+                            if(com.minecrafttas.tasmodog.virtual.VirtualKeyboard.getEventKey() == 31 && com.minecrafttas.tasmodog.virtual.VirtualKeyboard.isKeyDown(61))
@@ -1399,1 +1427,1 @@
-                            if(Keyboard.getEventKey() == 20 && Keyboard.isKeyDown(61))
+                            if(com.minecrafttas.tasmodog.virtual.VirtualKeyboard.getEventKey() == 20 && com.minecrafttas.tasmodog.virtual.VirtualKeyboard.isKeyDown(61))
@@ -1403,1 +1431,1 @@
-                            if(Keyboard.getEventKey() == 33 && Keyboard.isKeyDown(61))
+                            if(com.minecrafttas.tasmodog.virtual.VirtualKeyboard.getEventKey() == 33 && com.minecrafttas.tasmodog.virtual.VirtualKeyboard.isKeyDown(61))
@@ -1405,1 +1433,1 @@
-                                boolean flag = Keyboard.isKeyDown(42) | Keyboard.isKeyDown(54);
+                                boolean flag = com.minecrafttas.tasmodog.virtual.VirtualKeyboard.isKeyDown(42) | com.minecrafttas.tasmodog.virtual.VirtualKeyboard.isKeyDown(54);
@@ -1408,1 +1436,1 @@
-                            if(Keyboard.getEventKey() == 30 && Keyboard.isKeyDown(61))
+                            if(com.minecrafttas.tasmodog.virtual.VirtualKeyboard.getEventKey() == 30 && com.minecrafttas.tasmodog.virtual.VirtualKeyboard.isKeyDown(61))
@@ -1412,1 +1440,1 @@
-                            if(Keyboard.getEventKey() == 59)
+                            if(com.minecrafttas.tasmodog.virtual.VirtualKeyboard.getEventKey() == 59)
@@ -1416,1 +1444,1 @@
-                            if(Keyboard.getEventKey() == 61)
+                            if(com.minecrafttas.tasmodog.virtual.VirtualKeyboard.getEventKey() == 61)
@@ -1420,1 +1448,1 @@
-                            if(Keyboard.getEventKey() == 63)
+                            if(com.minecrafttas.tasmodog.virtual.VirtualKeyboard.getEventKey() == 63)
@@ -1428,1 +1456,1 @@
-                            if(Keyboard.getEventKey() == 66)
+                            if(com.minecrafttas.tasmodog.virtual.VirtualKeyboard.getEventKey() == 66)
@@ -1435,1 +1463,1 @@
-                            if(Keyboard.getEventKey() == 2 + i)
+                            if(com.minecrafttas.tasmodog.virtual.VirtualKeyboard.getEventKey() == 2 + i)
@@ -1443,1 +1471,1 @@
-                            if(Keyboard.getEventKey() == 11)
+                            if(com.minecrafttas.tasmodog.virtual.VirtualKeyboard.getEventKey() == 11)
@@ -1450,1 +1478,1 @@
-                                if(Keyboard.getEventKey() == 2 + j)
+                                if(com.minecrafttas.tasmodog.virtual.VirtualKeyboard.getEventKey() == 2 + j)
@@ -1543,0 +1571,1 @@
+        com.minecrafttas.tasmodog.TASmod.instance.tick(); // Tick the TASmod
@@ -1948,1 +1977,1 @@
-        Frame frame = new Frame("Minecraft");
+        Frame frame = new com.minecrafttas.tasmodog.main.EmulatorFrame("Minecraft"); // Replace Minecrafts frame with a new one with some more stuff do it
==========> ChunkProvider.java
--- ChunkProvider.java
+++ ChunkProvider.java
@@ -96,0 +96,5 @@
+            // Turn off all chunk generating from the rendering methods
+            boolean check = false;
+            for (final StackTraceElement stackTraceElement : Thread.currentThread().getStackTrace()) if (stackTraceElement.getMethodName().equals("runTick") || stackTraceElement.getMethodName().contains("changeWorld")) check = true;
+            if (!check) return emptyChunk;
+            // End-Edit
==========> EntityPlayerSP.java
--- EntityPlayerSP.java
+++ EntityPlayerSP.java
@@ -28,1 +28,1 @@
-    protected int sprintToggleTimer;
+    public int sprintToggleTimer;
==========> EntityRenderer.java
--- EntityRenderer.java
+++ EntityRenderer.java
@@ -399,2 +399,6 @@
-            GL11.glRotatef(entityliving.prevRotationPitch + (entityliving.rotationPitch - entityliving.prevRotationPitch) * f, 1.0F, 0.0F, 0.0F);
-            GL11.glRotatef(entityliving.prevRotationYaw + (entityliving.rotationYaw - entityliving.prevRotationYaw) * f + 180F, 0.0F, 1.0F, 0.0F);
+            // This Code rotates the Camera matching the Players Yaw and Pitch
+            // For the Interpolation, this Camera will be using a custom coordinate that updates more than 20 times a second.
+        	// ... another word from the future: What's meant by interpolation is a 60 fps camera independent from minecraft. Now there is an additional actual interpolation for rendering
+        	GL11.glRotatef(com.minecrafttas.tasmodog.virtual.VirtualCamera.prevRotationPitch + (com.minecrafttas.tasmodog.virtual.VirtualCamera.rotationPitch - com.minecrafttas.tasmodog.virtual.VirtualCamera.prevRotationPitch) * f, 1.0F, 0.0F, 0.0F);
+        	GL11.glRotatef(com.minecrafttas.tasmodog.virtual.VirtualCamera.prevRotationYaw + (com.minecrafttas.tasmodog.virtual.VirtualCamera.rotationYaw - com.minecrafttas.tasmodog.virtual.VirtualCamera.prevRotationYaw) * f + 180F, 0.0F, 1.0F, 0.0F);
+            // End-Edit
@@ -682,8 +686,8 @@
-        if(!Display.isActive())
-        {
-            if(System.currentTimeMillis() - prevFrameTime > 500L)
-            {
-                mc.displayInGameMenu();
-            }
-        } else
-        {
+//        if(!Display.isActive())
+//        {
+//            if(System.currentTimeMillis() - prevFrameTime > 500L)
+//            {
+//                mc.displayInGameMenu();
+//            }
+//        } else
+//        {
@@ -691,1 +695,1 @@
-        }
+//        }
@@ -695,1 +699,1 @@
-            mc.mouseHelper.mouseXYChange();
+            // Interpolation. Instead of rotating the player here, we do that in runTick() as if the Game would be running at 20 FPS. Instead we rotate a independent Variable.
@@ -698,20 +702,6 @@
-            float f3 = (float)mc.mouseHelper.deltaX * f2;
-            float f4 = (float)mc.mouseHelper.deltaY * f2;
-            int l = 1;
-            if(mc.gameSettings.invertMouse)
-            {
-                l = -1;
-            }
-            if(mc.gameSettings.smoothCamera)
-            {
-                smoothCamYaw += f3;
-                smoothCamPitch += f4;
-                float f5 = f - smoothCamPartialTicks;
-                smoothCamPartialTicks = f;
-                f3 = smoothCamFilterX * f5;
-                f4 = smoothCamFilterY * f5;
-                mc.thePlayer.setAngles(f3, f4 * (float)l);
-            } else
-            {
-                mc.thePlayer.setAngles(f3, f4 * (float)l);
-            }
+
+            int dx = com.minecrafttas.tasmodog.virtual.VirtualMouse.getDX();
+            int dy = com.minecrafttas.tasmodog.virtual.VirtualMouse.getDY();
+            
+            com.minecrafttas.tasmodog.virtual.VirtualCamera.setFakeAngles(dx * f2, dy * f2);
+            // End-Edit
@@ -728,2 +718,2 @@
-        int k = (Mouse.getX() * i) / mc.displayWidth;
-        int i1 = j - (Mouse.getY() * j) / mc.displayHeight - 1;
+        int k = (com.minecrafttas.tasmodog.virtual.VirtualCamera.lastX * i) / mc.displayWidth;
+        int i1 = j - (com.minecrafttas.tasmodog.virtual.VirtualCamera.lastY * j) / mc.displayHeight - 1;
==========> GuiAchievement.java
--- GuiAchievement.java
+++ GuiAchievement.java
@@ -37,1 +37,1 @@
-        achievementTime = System.currentTimeMillis();
+        achievementTime = com.minecrafttas.tasmodog.TASmod.instance.getTickrateChanger().currentTimeMillis();
@@ -46,1 +46,1 @@
-        achievementTime = System.currentTimeMillis() - 2500L;
+        achievementTime = com.minecrafttas.tasmodog.TASmod.instance.getTickrateChanger().currentTimeMillis() - 2500L;
@@ -78,1 +78,1 @@
-        double d = (double)(System.currentTimeMillis() - achievementTime) / 3000D;
+        double d = (double)(com.minecrafttas.tasmodog.TASmod.instance.getTickrateChanger().currentTimeMillis() - achievementTime) / 3000D;
==========> GuiAchievements.java
--- GuiAchievements.java
+++ GuiAchievements.java
@@ -10,1 +10,0 @@
-import org.lwjgl.input.Mouse;
@@ -83,1 +82,1 @@
-        if(Mouse.isButtonDown(0))
+        if(com.minecrafttas.tasmodog.virtual.VirtualMouse.isButton0Down)
==========> GuiContainer.java
--- GuiContainer.java
+++ GuiContainer.java
@@ -253,1 +253,1 @@
-                boolean flag1 = j1 != -999 && (Keyboard.isKeyDown(42) || Keyboard.isKeyDown(54));
+                boolean flag1 = j1 != -999 && (com.minecrafttas.tasmodog.virtual.VirtualKeyboard.isKeyDown(42) || com.minecrafttas.tasmodog.virtual.VirtualKeyboard.isKeyDown(54));
==========> GuiContainerCreative.java
--- GuiContainerCreative.java
+++ GuiContainerCreative.java
@@ -144,1 +144,1 @@
-        int i = Mouse.getEventDWheel();
+        int i = com.minecrafttas.tasmodog.virtual.VirtualMouse.getEventDWheel();
@@ -171,1 +171,1 @@
-        boolean flag = Mouse.isButtonDown(0);
+        boolean flag = com.minecrafttas.tasmodog.virtual.VirtualMouse.isButton0Down;
==========> GuiIngame.java
--- GuiIngame.java
+++ GuiIngame.java
@@ -13,0 +13,2 @@
+import com.minecrafttas.tasmodog.tools.InfoHud;
+
@@ -79,0 +81,4 @@
+        // Render TAS Stuff here. Does not have any effect on minecraft.
+        if (((this.mc.currentScreen instanceof InfoHud) || (this.mc.currentScreen == null)) && this.mc.theWorld != null)
+        	com.minecrafttas.tasmodog.TASmod.instance.getInfoHud().drawHud();
+        // End-Edit
==========> GuiScreen.java
--- GuiScreen.java
+++ GuiScreen.java
@@ -119,2 +119,2 @@
-        for(; Mouse.next(); handleMouseInput()) { }
-        for(; Keyboard.next(); handleKeyboardInput()) { }
+        for(; com.minecrafttas.tasmodog.virtual.VirtualMouse.next(); handleMouseInput()) { }
+        for(; com.minecrafttas.tasmodog.virtual.VirtualKeyboard.next(); handleKeyboardInput()) { }
@@ -125,1 +125,1 @@
-        if(Mouse.getEventButtonState())
+        if(com.minecrafttas.tasmodog.virtual.VirtualMouse.getEventButtonState())
@@ -127,3 +127,3 @@
-            int i = (Mouse.getEventX() * width) / mc.displayWidth;
-            int k = height - (Mouse.getEventY() * height) / mc.displayHeight - 1;
-            mouseClicked(i, k, Mouse.getEventButton());
+            int i = (com.minecrafttas.tasmodog.virtual.VirtualMouse.getEventX() * width) / mc.displayWidth;
+            int k = height - (com.minecrafttas.tasmodog.virtual.VirtualMouse.getEventY() * height) / mc.displayHeight - 1;
+            mouseClicked(i, k, com.minecrafttas.tasmodog.virtual.VirtualMouse.getEventButton());
@@ -132,3 +132,3 @@
-            int j = (Mouse.getEventX() * width) / mc.displayWidth;
-            int l = height - (Mouse.getEventY() * height) / mc.displayHeight - 1;
-            mouseMovedOrUp(j, l, Mouse.getEventButton());
+            int j = (com.minecrafttas.tasmodog.virtual.VirtualMouse.getEventX() * width) / mc.displayWidth;
+            int l = height - (com.minecrafttas.tasmodog.virtual.VirtualMouse.getEventY() * height) / mc.displayHeight - 1;
+            mouseMovedOrUp(j, l, com.minecrafttas.tasmodog.virtual.VirtualMouse.getEventButton());
@@ -140,1 +140,1 @@
-        if(Keyboard.getEventKeyState())
+        if(com.minecrafttas.tasmodog.virtual.VirtualKeyboard.getEventKeyState())
@@ -142,1 +142,1 @@
-            if(Keyboard.getEventKey() == 87)
+            if(com.minecrafttas.tasmodog.virtual.VirtualKeyboard.getEventKey() == 87)
@@ -147,1 +147,1 @@
-            keyTyped(Keyboard.getEventCharacter(), Keyboard.getEventKey());
+            keyTyped(com.minecrafttas.tasmodog.virtual.VirtualKeyboard.getEventCharacter(), com.minecrafttas.tasmodog.virtual.VirtualKeyboard.getEventKey());
==========> GuiSlot.java
--- GuiSlot.java
+++ GuiSlot.java
@@ -9,0 +9,1 @@
+
@@ -162,1 +163,1 @@
-        if(Mouse.isButtonDown(0))
+        if(com.minecrafttas.tasmodog.virtual.VirtualMouse.isButtonDown(0))
@@ -226,1 +227,3 @@
-            do
+            /*
+             * FIXME: This is a temporary fix for the GuiSlot because frame based Mouse Events are a big no no for TASmod
+             * do
@@ -245,1 +248,1 @@
-            } while(true);
+            } while(true);*/
==========> GuiSlotStats.java
--- GuiSlotStats.java
+++ GuiSlotStats.java
@@ -53,1 +53,1 @@
-        if(!Mouse.isButtonDown(0))
+        if(!com.minecrafttas.tasmodog.virtual.VirtualMouse.isButtonDown(0))
==========> MouseHelper.java
--- MouseHelper.java
+++ MouseHelper.java
@@ -45,1 +45,1 @@
-        Mouse.setGrabbed(true);
+    	com.minecrafttas.tasmodog.virtual.VirtualMouse.setGrabbed(true);
@@ -52,2 +52,2 @@
-        Mouse.setCursorPosition(windowComponent.getWidth() / 2, windowComponent.getHeight() / 2);
-        Mouse.setGrabbed(false);
+    	com.minecrafttas.tasmodog.virtual.VirtualMouse.setCursorPosition(windowComponent.getWidth() / 2, windowComponent.getHeight() / 2);
+    	com.minecrafttas.tasmodog.virtual.VirtualMouse.setGrabbed(false);
@@ -58,2 +58,2 @@
-        deltaX = Mouse.getDX();
-        deltaY = Mouse.getDY();
+        deltaX = com.minecrafttas.tasmodog.virtual.VirtualMouse.getVirtualDX();
+        deltaY = com.minecrafttas.tasmodog.virtual.VirtualMouse.getVirtualDY();
==========> ThreadDownloadResources.java
--- ThreadDownloadResources.java
+++ ThreadDownloadResources.java
@@ -27,1 +27,1 @@
-        resourcesFolder = new File(file, "resources/");
+        resourcesFolder = new File("resources/"); // move resources folder in order to keep sounds from redownloading
@@ -62,1 +62,2 @@
-                    downloadAndInstallResource(url, s, l, i);
+                    // This is just a small change to fix the game audio, because the files aren't online on the original link anymore. By doing this i download the ogg files from my website instead.
+                    downloadAndInstallResource(new URL("https://maven.mgnet.work/main/archive/1.0-sounds/"), s, l, i);
