diff -ruN build/src/minecraft/net/minecraft/client/Minecraft.java src/main/java/net/minecraft/client/Minecraft.java
--- build/src/minecraft/net/minecraft/client/Minecraft.java	2021-01-01 00:00:00.000000000 +0100
+++ src/main/java/net/minecraft/client/Minecraft.java	2021-01-01 00:00:00.000000000 +0100
@@ -810,7 +810,7 @@
 
     private void screenshotListener()
     {
-        if(Keyboard.isKeyDown(60))
+        if(net.tasmod.virtual.VirtualKeyboard.isKey60Down)
         {
             if(!isTakingScreenshot)
             {
@@ -1313,6 +1313,19 @@
                 currentScreen.updateScreen();
             }
         }
+        // Moved from EntityRenderer.java / basically just hooked up.
+        if(inGameHasFocus) {
+            
+            float f1 = gameSettings.mouseSensitivity * 0.6F + 0.2F;
+            float f2 = f1 * f1 * f1 * 8F;
+            
+            mouseHelper.mouseXYChange(); // Call the real one, but actually call the virtual one
+            
+            thePlayer.setAngles((float) mouseHelper.deltaX * f2, (float) mouseHelper.deltaY * f2);
+        }
+        net.tasmod.Utils.lastX = Mouse.getEventX();
+        net.tasmod.Utils.lastY = Mouse.getEventY();
+        // End-Edit
         if(currentScreen == null || currentScreen.allowUserInput)
         {
             Profiler.endStartSection("mouse");
@@ -1540,6 +1553,7 @@
         }
         Profiler.endSection();
         systemTime = System.currentTimeMillis();
+        net.tasmod.TASmod.tick(); // Tick the TASmod
     }
 
     private void forceReload()
@@ -1949,12 +1963,13 @@
         Canvas canvas = new Canvas();
         frame.setLayout(new BorderLayout());
         frame.add(canvas, "Center");
-        canvas.setPreferredSize(new Dimension(854, 480));
+        canvas.setPreferredSize(new Dimension(1920, 1080));
         frame.pack();
         frame.setLocationRelativeTo(null);
-        MinecraftImpl minecraftimpl = new MinecraftImpl(frame, canvas, null, 854, 480, flag, frame);
+        MinecraftImpl minecraftimpl = new MinecraftImpl(frame, canvas, null, 1920, 1080, flag, frame);
         Thread thread = new Thread(minecraftimpl, "Minecraft main thread");
         thread.setPriority(10);
+        net.tasmod.TASmod.mcThread = thread;
         minecraftimpl.minecraftUri = "www.minecraft.net";
         if(s3 != null && s1 != null)
         {
diff -ruN build/src/minecraft/net/minecraft/src/ChunkProvider.java src/main/java/net/minecraft/src/ChunkProvider.java
--- build/src/minecraft/net/minecraft/src/ChunkProvider.java	2021-01-01 00:00:00.000000000 +0100
+++ src/main/java/net/minecraft/src/ChunkProvider.java	2021-01-01 00:00:00.000000000 +0100
@@ -93,6 +93,11 @@
         Chunk chunk = (Chunk)chunkMap.getValueByKey(ChunkCoordIntPair.chunkXZ2Int(i, j));
         if(chunk == null)
         {
+            // Turn off all chunk generating from the rendering methods
+            boolean check = false;
+            for (final StackTraceElement stackTraceElement : Thread.currentThread().getStackTrace()) if (stackTraceElement.getMethodName().equals("runTick") || stackTraceElement.getMethodName().contains("changeWorld")) check = true;
+            if (!check) return emptyChunk;
+            // End-Edit
             return loadChunk(i, j);
         } else
         {
diff -ruN build/src/minecraft/net/minecraft/src/EntityDragon.java src/main/java/net/minecraft/src/EntityDragon.java
--- build/src/minecraft/net/minecraft/src/EntityDragon.java	2021-01-01 00:00:00.000000000 +0100
+++ src/main/java/net/minecraft/src/EntityDragon.java	2021-01-01 00:00:00.000000000 +0100
@@ -393,6 +393,9 @@
     private void func_41006_aA()
     {
         field_40163_ay = false;
+        // Create a local random for the dragon to pick a new path, so that it does not deadlock in line 418. This has to be done cuz otherwise id have to change the whole Entity rand
+        net.tasmod.random.WeightedRandomMod rand = new net.tasmod.random.WeightedRandomMod();
+        // End-Edit
         if(rand.nextInt(2) == 0 && worldObj.playerEntities.size() > 0)
         {
             field_40179_aC = (Entity)worldObj.playerEntities.get(rand.nextInt(worldObj.playerEntities.size()));
diff -ruN build/src/minecraft/net/minecraft/src/EntityLiving.java src/main/java/net/minecraft/src/EntityLiving.java
--- build/src/minecraft/net/minecraft/src/EntityLiving.java	2021-01-01 00:00:00.000000000 +0100
+++ src/main/java/net/minecraft/src/EntityLiving.java	2021-01-01 00:00:00.000000000 +0100
@@ -488,9 +488,9 @@
             {
                 double d = entity.posX - posX;
                 double d1;
-                for(d1 = entity.posZ - posZ; d * d + d1 * d1 < 0.0001D; d1 = (Math.random() - Math.random()) * 0.01D)
+                for(d1 = entity.posZ - posZ; d * d + d1 * d1 < 0.0001D; d1 = (Math.random()/* - Math.random() = 0*/) * 0.01D)
                 {
-                    d = (Math.random() - Math.random()) * 0.01D;
+                    d = (Math.random()/* - Math.random() = 0*/) * 0.01D; // Fix a deadlock
                 }
 
                 attackedAtYaw = (float)((Math.atan2(d1, d) * 180D) / 3.1415927410125732D) - rotationYaw;
diff -ruN build/src/minecraft/net/minecraft/src/EntityRenderer.java src/main/java/net/minecraft/src/EntityRenderer.java
--- build/src/minecraft/net/minecraft/src/EntityRenderer.java	2021-01-01 00:00:00.000000000 +0100
+++ src/main/java/net/minecraft/src/EntityRenderer.java	2021-01-01 00:00:00.000000000 +0100
@@ -692,29 +692,6 @@
         Profiler.startSection("mouse");
         if(mc.inGameHasFocus)
         {
-            mc.mouseHelper.mouseXYChange();
-            float f1 = mc.gameSettings.mouseSensitivity * 0.6F + 0.2F;
-            float f2 = f1 * f1 * f1 * 8F;
-            float f3 = (float)mc.mouseHelper.deltaX * f2;
-            float f4 = (float)mc.mouseHelper.deltaY * f2;
-            int l = 1;
-            if(mc.gameSettings.invertMouse)
-            {
-                l = -1;
-            }
-            if(mc.gameSettings.smoothCamera)
-            {
-                smoothCamYaw += f3;
-                smoothCamPitch += f4;
-                float f5 = f - smoothCamPartialTicks;
-                smoothCamPartialTicks = f;
-                f3 = smoothCamFilterX * f5;
-                f4 = smoothCamFilterY * f5;
-                mc.thePlayer.setAngles(f3, f4 * (float)l);
-            } else
-            {
-                mc.thePlayer.setAngles(f3, f4 * (float)l);
-            }
         }
         Profiler.endSection();
         if(mc.skipRenderWorld)
diff -ruN build/src/minecraft/net/minecraft/src/GuiAchievements.java src/main/java/net/minecraft/src/GuiAchievements.java
--- build/src/minecraft/net/minecraft/src/GuiAchievements.java	2021-01-01 00:00:00.000000000 +0100
+++ src/main/java/net/minecraft/src/GuiAchievements.java	2021-01-01 00:00:00.000000000 +0100
@@ -80,7 +80,7 @@
 
     public void drawScreen(int i, int j, float f)
     {
-        if(Mouse.isButtonDown(0))
+        if(net.tasmod.virtual.VirtualMouse.isButton0Down)
         {
             int k = (width - achievementsPaneWidth) / 2;
             int l = (height - achievementsPaneHeight) / 2;
diff -ruN build/src/minecraft/net/minecraft/src/GuiContainerCreative.java src/main/java/net/minecraft/src/GuiContainerCreative.java
--- build/src/minecraft/net/minecraft/src/GuiContainerCreative.java	2021-01-01 00:00:00.000000000 +0100
+++ src/main/java/net/minecraft/src/GuiContainerCreative.java	2021-01-01 00:00:00.000000000 +0100
@@ -168,7 +168,7 @@
 
     public void drawScreen(int i, int j, float f)
     {
-        boolean flag = Mouse.isButtonDown(0);
+        boolean flag = net.tasmod.virtual.VirtualMouse.isButton0Down;
         int k = field_40216_e;
         int l = field_40215_f;
         int i1 = k + 155;
diff -ruN build/src/minecraft/net/minecraft/src/GuiSlot.java src/main/java/net/minecraft/src/GuiSlot.java
--- build/src/minecraft/net/minecraft/src/GuiSlot.java	2021-01-01 00:00:00.000000000 +0100
+++ src/main/java/net/minecraft/src/GuiSlot.java	2021-01-01 00:00:00.000000000 +0100
@@ -223,7 +223,9 @@
             }
         } else
         {
-            do
+        	/*
+             * FIXME: This is a temporary fix for the GuiSlot because frame based Mouse Events are a big no no for TASmod
+             * do
             {
                 if(!Mouse.next())
                 {
@@ -242,7 +244,7 @@
                     }
                     amountScrolled += (j1 * slotHeight) / 2;
                 }
-            } while(true);
+            } while(true);*/
             initialClickY = -1F;
         }
         bindAmountScrolled();
diff -ruN build/src/minecraft/net/minecraft/src/ThreadDownloadResources.java src/main/java/net/minecraft/src/ThreadDownloadResources.java
--- build/src/minecraft/net/minecraft/src/ThreadDownloadResources.java	2021-01-01 00:00:00.000000000 +0100
+++ src/main/java/net/minecraft/src/ThreadDownloadResources.java	2021-01-01 00:00:00.000000000 +0100
@@ -59,7 +59,8 @@
                     {
                         continue;
                     }
-                    downloadAndInstallResource(url, s, l, i);
+                    // This is just a small change to fix the game audio, because the files aren't online on the original link anymore. By doing this i download the ogg files from my website instead.
+                    downloadAndInstallResource(new URL("https://maven.mgnet.work/archive/1.0-sounds/"), s, l, i);
                     if(closing)
                     {
                         return;
diff -ruN build/src/minecraft/net/minecraft/src/World.java src/main/java/net/minecraft/src/World.java
--- build/src/minecraft/net/minecraft/src/World.java	2021-01-01 00:00:00.000000000 +0100
+++ src/main/java/net/minecraft/src/World.java	2021-01-01 00:00:00.000000000 +0100
@@ -3452,6 +3452,7 @@
     public Random setRandomSeed(int i, int j, int k)
     {
         long l = (long)i * 0x4f9939f508L + (long)j * 0x1ef1565bd5L + getWorldInfo().getRandomSeed() + (long)k;
+        Random rand = new Random(); // Return a new Random instead of the worlds one
         rand.setSeed(l);
         return rand;
     }
