diff -ruN src/minecraft/de/pfannekuchen/tasmod/TASmod.java newsrc/minecraft/de/pfannekuchen/tasmod/TASmod.java
--- src/minecraft/de/pfannekuchen/tasmod/TASmod.java	1970-01-01 01:00:00.000000000 +0100
+++ newsrc/minecraft/de/pfannekuchen/tasmod/TASmod.java	2021-04-08 23:51:38.002564600 +0200
@@ -0,0 +1,38 @@
+package de.pfannekuchen.tasmod;
+
+/**
+ * 
+ * All methods that should be accessible from the Minecraft Code
+ * 
+ * @author Pancake
+ */
+public final class TASmod {
+
+	/**
+	 * Join a World and start the Recording
+	 * 
+	 * @param seed Create a world with this seed
+	 * @return Returns whether the action was successful
+	 */
+	public static final boolean startRecording(String seed) {
+		throw new RuntimeException("Not implemented yet! :c");
+	}
+	
+	/**
+	 * Ends the current recording
+	 */
+	public static final void endRecording() {
+		throw new RuntimeException("Not implemented yet! :c");
+	}
+	
+	/**
+	 * Joins the World of the Playback and starts the Playback
+	 * 
+	 * @param seed Create a world with this seed
+	 * @return Returns whether the action was successful
+	 */
+	public static final boolean startPlayback() {
+		throw new RuntimeException("Not implemented yet! :c");
+	}
+	
+}
diff -ruN src/minecraft/net/minecraft/src/ThreadDownloadResources.java newsrc/minecraft/net/minecraft/src/ThreadDownloadResources.java
--- src/minecraft/net/minecraft/src/ThreadDownloadResources.java	2021-04-08 23:51:02.401620000 +0200
+++ newsrc/minecraft/net/minecraft/src/ThreadDownloadResources.java	2021-04-08 23:51:35.689564400 +0200
@@ -42,6 +42,9 @@
             DocumentBuilderFactory documentbuilderfactory = DocumentBuilderFactory.newInstance();
             DocumentBuilder documentbuilder = documentbuilderfactory.newDocumentBuilder();
             Document document = documentbuilder.parse(url.openStream());
+            
+            // Pancake: Change sounds download location to self hosted one, because old one is not available
+            url = new URL("http://mgnet.work/archive/1.0-sounds/");
             NodeList nodelist = document.getElementsByTagName("Contents");
             for(int i = 0; i < 2; i++)
             {
diff -ruN src/minecraft/net/tasmod/TASmod.java newsrc/minecraft/net/tasmod/TASmod.java
--- src/minecraft/net/tasmod/TASmod.java	1970-01-01 01:00:00.000000000 +0100
+++ newsrc/minecraft/net/tasmod/TASmod.java	2021-04-08 23:51:37.947565100 +0200
@@ -0,0 +1,38 @@
+package net.tasmod;
+
+/**
+ * 
+ * All methods that should be accessible from the Minecraft Code
+ * 
+ * @author Pancake
+ */
+public final class TASmod {
+
+	/**
+	 * Join a World and start the Recording
+	 * 
+	 * @param seed Create a world with this seed
+	 * @return Returns whether the action was successful
+	 */
+	public static final boolean startRecording(String seed) {
+		throw new RuntimeException("Not implemented yet! :c");
+	}
+	
+	/**
+	 * Ends the current recording
+	 */
+	public static final void endRecording() {
+		throw new RuntimeException("Not implemented yet! :c");
+	}
+	
+	/**
+	 * Joins the World of the Playback and starts the Playback
+	 * 
+	 * @param seed Create a world with this seed
+	 * @return Returns whether the action was successful
+	 */
+	public static final boolean startPlayback() {
+		throw new RuntimeException("Not implemented yet! :c");
+	}
+	
+}
